libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.m

```diff
@@ -0,0 +1,827 @@
+        return;
+    }
+    [SFSDKCoreLogger i:[self class] format:@"%@: Restarting in-progress authentication process.", NSStringFromSelector(_cmd)];
+    [self.coordinator stopAuthentication];
+    [self refreshCredentials];
+}
+
+#pragma mark - private class members
++ (BOOL)errorIsInvalidAuthCredentials:(NSError *)error
+{
+    BOOL errorIsInvalidCreds = NO;
+    if (error.domain == kSFOAuthErrorDomain) {
+        if (error.code == kSFOAuthErrorInvalidGrant) {
+            errorIsInvalidCreds = YES;
+        }
+    }
+    return errorIsInvalidCreds;
+}
+
+/**
+ * Evaluates an NSError object to see if it represents a network failure during
+ * an attempted connection.
+ * @param error The NSError to evaluate.
+ * @return YES if the error represents a network failure, NO otherwise.
+ */
++ (BOOL)errorIsNetworkFailure:(NSError *)error
+{
+    BOOL isNetworkFailure = NO;
+
+    if (error == nil || error.domain == nil)
+        return isNetworkFailure;
+
+    if ([error.domain isEqualToString:NSURLErrorDomain]) {
+        switch (error.code) {
+            case NSURLErrorTimedOut:
+            case NSURLErrorCannotConnectToHost:
+            case NSURLErrorNetworkConnectionLost:
+            case NSURLErrorNotConnectedToInternet:
+            case NSURLErrorInternationalRoamingOff:
+                isNetworkFailure = YES;
+                break;
+            default:
+                break;
+        }
+    } else if ([error.domain isEqualToString:kSFOAuthErrorDomain]) {
+        switch (error.code) {
+            case kSFOAuthErrorTimeout:
+                isNetworkFailure = YES;
+                break;
+            default:
+                break;
+        }
+    }
+
+    return isNetworkFailure;
+}
+
+#pragma mark - SFSDKOAuthClientProvider
++ (SFSDKOAuthClient *)idpAuthInstance:(SFSDKOAuthClientConfig *)config {
+    if (self.clientProvider) {
+        return [self.clientProvider idpAuthInstance:config];
+    }
+
+    static dispatch_once_t once = 0;
+    dispatch_once(&once, ^{
+        _idpInstance = [[self alloc] init];
+    });
+    return _idpInstance;
+
+}
+
++ (SFSDKOAuthClient *)nativeBrowserAuthInstance:(SFSDKOAuthClientConfig *)config {
+    if (self.clientProvider) {
+        return [self.clientProvider nativeBrowserAuthInstance:config];
+    }
+
+    static dispatch_once_t once = 0;
+    dispatch_once(&once, ^{
+        _sharedInstance = [[self alloc] initWithConfig:config];
+    });
+    return _sharedInstance;
+
+}
+
++ (SFSDKOAuthClient *)webviewAuthInstance:(SFSDKOAuthClientConfig *)config {
+    if (self.clientProvider) {
+        return [self.clientProvider webviewAuthInstance:config];
+    }
+    return [[self alloc] initWithConfig:config];
+}
+
++ (SFSDKOAuthClient *)clientWithCredentials:(SFOAuthCredentials *)credentials configBlock:(void(^)(SFSDKOAuthClientConfig *))configBlock {
+
+    SFSDKOAuthClientConfig *config = [SFSDKOAuthClientConfig new];
```

Comment:
- `new` is kind of old—sorry for the pun. 😉  `alloc/init` is generally the convention these days.
- Se should be consistent, at any rate.  We're mixing both instantiation approaches in the same method.